### PR TITLE
Feature/877 keep account alive

### LINKED ISSUE #877: UX education - decimals

### PR DESCRIPTION
**Pull request by Ekaanth**
fix #877 

blocked by https://github.com/litentry/litentry-graph/pull/188

Simulator Screen Shot - iPhone 12 - 2022-03-16 at 14 31 19

Simulator Screen Shot - iPhone 12 - 2022-03-16 at 14 46 15

### CODE REVIEW

File: src/ui/screens/SendFundScreen.tsx [hunk 1]
```diff
@@ -30,11 +34,20 @@ export function SendFundScreen({navigation, route}: Props) {
...
   const {data: chainInfo} = useChainInfo();
+  const [isKeepAlive, setisKeepAlive] = React.useState(true);
+  const {currentNetwork} = useContext(NetworkContext);
+  const snackbar = useSnackbar();
 
   useEffect(() => {
     ref.current?.open();
   }, []);
 
+  const isAccountValid = useMemo(() => {
```
Comment (silva-fj) [inline src/ui/screens/SendFundScreen.tsx]:
```suggestion
  const isToAddressValid = useMemo(() => {
```

File: src/ui/screens/SendFundScreen.tsx [hunk 2]
```diff
@@ -30,11 +34,20 @@ export function SendFundScreen({navigation, route}: Props) {
...
+    return to ? isAddressValid(currentNetwork, to) : false;
```
Comment (silva-fj) [inline src/ui/screens/SendFundScreen.tsx]: Let's change this name `toAddress` instead of just `to`.

File: src/ui/screens/SendFundScreen.tsx [hunk 3]
```diff
@@ -57,17 +70,40 @@ export function SendFundScreen({navigation, route}: Props) {
...
+              helperText="Scan a contact address or paste the address you want to send funds to."
+            />
             <TextInput
               autoComplete="off"
               placeholder="To"
               value={to}
               onChangeText={(nextValue) => setTo(nextValue)}
               right={<TextInput.Icon name="qrcode" onPress={() => setScanning(true)} />}
             />
+            {!isAccountValid && to ? <HelperText type={'error'}>Enter a valid address</HelperText> : null}
```
Comment (silva-fj) [inline src/ui/screens/SendFundScreen.tsx]:
```suggestion
            {!isToAddressValid ? <HelperText type={'error'}>Enter a valid address</HelperText> : null}
```

File: src/ui/screens/SendFundScreen.tsx [hunk 4]
```diff
@@ -30,11 +34,20 @@ export function SendFundScreen({navigation, route}: Props) {
   const [scanning, setScanning] = React.useState(false);
   const startTx = useApiTx();
   const {data: chainInfo} = useChainInfo();
+  const [isKeepAlive, setisKeepAlive] = React.useState(true);
```
Comment (silva-fj) [inline src/ui/screens/SendFundScreen.tsx]: NIT: `isKeepAliveEnabled` or `isKeepAliveActive`

File: src/ui/screens/SendFundScreen.tsx [hunk 5]
```diff
@@ -76,22 +112,25 @@ export function SendFundScreen({navigation, route}: Props) {
                     const _amountBN = stringToBn(chainInfo.registry, amount);
                     startTx({
                       address,
-                      txMethod: 'balances.transferKeepAlive',
+                      txMethod: `${isKeepAlive ? `balances.transferKeepAlive` : `balances.transferKeepAlive`}`,
```
Comment (silva-fj) [inline src/ui/screens/SendFundScreen.tsx]:
```suggestion
                      txMethod: `${isKeepAlive ? `balances.transferKeepAlive` : `balances.transfer`}`,
```

File: src/ui/screens/SendFundScreen.tsx [hunk 6]
```diff
@@ -76,22 +112,25 @@ export function SendFundScreen({navigation, route}: Props) {
...
                       params: [to, _amountBN],
                     })
                       .then(() => {
+                        snackbar('Funds transfered');
                         navigation.goBack();
                       })
                       .catch((e: Error) => {
                         console.warn(e);
```
Comment (silva-fj) [inline src/ui/screens/SendFundScreen.tsx]: If this fails we should inform the user that something went wrong.

File: src/ui/screens/SendFundScreen.tsx [hunk 7]
```diff
@@ -57,17 +70,40 @@ export function SendFundScreen({navigation, route}: Props) {
...
             <TextInput
               autoComplete="off"
               placeholder="To"
-              value={to}
-              onChangeText={(nextValue) => setTo(nextValue)}
+              value={toAddress}
+              onChangeText={(nextValue) => setToAddress(nextValue)}
               right={<TextInput.Icon name="qrcode" onPress={() => setScanning(true)} />}
             />
+            {!isToAddressValid ? <HelperText type={'error'}>Enter a valid address</HelperText> : null}
```
Comment (ishrak-imam) [inline src/ui/screens/SendFundScreen.tsx]: NIT 💅 
```suggestion
            {!isToAddressValid ? <HelperText type="error">Enter a valid address</HelperText> : null}
```

And for other cases also.

File: src/ui/screens/SendFundScreen.tsx [hunk 8]
```diff
@@ -133,6 +172,15 @@ const styles = StyleSheet.create({
     alignItems: 'center',
     borderRadius: 10,
   },
+  keepAlive: {
+    flexDirection: 'row',
+    justifyContent: 'flex-end',
+    paddingVertical: 15,
```
Comment (ishrak-imam) [inline src/ui/screens/SendFundScreen.tsx]: NIT 💅 

We could use the standard padding or a multiple of that.

```suggestion
    paddingVertical: standardPadding * 2,
```

File: src/ui/screens/SendFundScreen.tsx [hunk 9]
```diff
@@ -133,6 +172,15 @@ const styles = StyleSheet.create({
...
+  },
+  keepAliveContainer: {
+    paddingVertical: 5,
+    paddingHorizontal: 10,
```
Comment (ishrak-imam) [inline src/ui/screens/SendFundScreen.tsx]:
```suggestion
    paddingHorizontal: standardPadding,
```

**Review by ishrak-imam — APPROVED** [review]
LGTM !!!

Awesome 🎉

**Review by silva-fj — APPROVED** [review]

**Comment by Ekaanth** [comment]
@franjs @ishrak-imam  I have an additional requirement here, I believe, to validate the form when the user is set the toggler to `keepAlive` the entered value should be one less than the `formattedfree` correct me if I'm wrong here.

**Comment by ishrak-imam** [comment]
> @franjs @ishrak-imam I have an additional requirement here, I believe, to validate the form when the user is set the toggler to `keepAlive` the entered value should be one less than the `formattedfree` correct me if I'm wrong here.

Yes, you are correct.

**Comment by Ekaanth** [comment]
> > @franjs @ishrak-imam I have an additional requirement here, I believe, to validate the form when the user is set the toggler to `keepAlive` the entered value should be one less than the `formattedfree` correct me if I'm wrong here.
> 
> Yes, you are correct.

@ishrak-imam  Thanks for confirming. In that case when the stay alive is active what is the maximum token that can be transferred. how do I handle it?

**Comment by ishrak-imam** [comment]
> > > @franjs @ishrak-imam I have an additional requirement here, I believe, to validate the form when the user is set the toggler to `keepAlive` the entered value should be one less than the `formattedfree` correct me if I'm wrong here.
> > 
> > 
> > Yes, you are correct.
> 
> @ishrak-imam Thanks for confirming. In that case when the stay alive is active what is the maximum token that can be transferred. how do I handle it?

We can pass the keepAlive configuration to the `<BalanceInput />` component and do the checking there. Also the `<BalanceInput />` component should actually return something like a `isValid` status along with the input amount.

If we think from one step back, The `<BalanceInput />` component can encapsulate the keepAlive toggle and all the logic in itself. It can only sent out the input amount and the `hasEnoughFunds` status to the parent.

**Comment by Ekaanth** [comment]
> > > > @franjs @ishrak-imam I have an additional requirement here, I believe, to validate the form when the user is set the toggler to `keepAlive` the entered value should be one less than the `formattedfree` correct me if I'm wrong here.
> > > 
> > > 
> > > Yes, you are correct.
> > 
> > 
> > @ishrak-imam Thanks for confirming. In that case when the stay alive is active what is the maximum token that can be transferred. how do I handle it?
> 
> We can pass the keepAlive configuration to the `<BalanceInput />` component and do the checking there. Also the `<BalanceInput />` component should actually return something like a `isValid` status along with the input amount.
> 
> If we think from one step back, The `<BalanceInput />` component can encapsulate the keepAlive toggle and all the logic in itself. It can only sent out the input amount and the `hasEnoughFunds` status to the parent.

In this case, we can send the bond value to the component as well. My question here is how do we determine the token's value? For example an account has 10DOT and they want to transfer keeping their account alive, it should be formatedFree-bondValue-fee right? how do we do this?

**Comment by silva-fj** [comment]
> > > > > @franjs @ishrak-imam I have an additional requirement here, I believe, to validate the form when the user is set the toggler to `keepAlive` the entered value should be one less than the `formattedfree` correct me if I'm wrong here.
> > > > 
> > > > 
> > > > Yes, you are correct.
> > > 
> > > 
> > > @ishrak-imam Thanks for confirming. In that case when the stay alive is active what is the maximum token that can be transferred. how do I handle it?
> > 
> > 
> > We can pass the keepAlive configuration to the `<BalanceInput />` component and do the checking there. Also the `<BalanceInput />` component should actually return something like a `isValid` status along with the input amount.
> > If we think from one step back, The `<BalanceInput />` component can encapsulate the keepAlive toggle and all the logic in itself. It can only sent out the input amount and the `hasEnoughFunds` status to the parent.
> 
> In this case, we can send the bond value to the component as well. My question here is how do we determine the token's value? For example an account has 10DOT and they want to transfer keeping their account alive, it should be formatedFree-bondValue-fee right? how do we do this?

You can take a look at the logic in the polkadotjs app [here](https://github.com/polkadot-js/apps/blob/master/packages/page-accounts/src/modals/Transfer.tsx)

**Comment by ishrak-imam** [comment]
> > > > > @franjs @ishrak-imam I have an additional requirement here, I believe, to validate the form when the user is set the toggler to `keepAlive` the entered value should be one less than the `formattedfree` correct me if I'm wrong here.
> > > > 
> > > > 
> > > > Yes, you are correct.
> > > 
> > > 
> > > @ishrak-imam Thanks for confirming. In that case when the stay alive is active what is the maximum token that can be transferred. how do I handle it?
> > 
> > 
> > We can pass the keepAlive configuration to the `<BalanceInput />` component and do the checking there. Also the `<BalanceInput />` component should actually return something like a `isValid` status along with the input amount.
> > If we think from one step back, The `<BalanceInput />` component can encapsulate the keepAlive toggle and all the logic in itself. It can only sent out the input amount and the `hasEnoughFunds` status to the parent.
> 
> In this case, we can send the bond value to the component as well. My question here is how do we determine the token's value? For example an account has 10DOT and they want to transfer keeping their account alive, it should be formatedFree-bondValue-fee right? how do we do this?

The only responsibility of the `BalanceInput` component is to check if the entered amount is less than or equal to the free balance of the selected account. And if keep alive is enabled entered amount should be less than or equal to freeBalance - existential deposit.

For the voting bond according to polkadot js app the keepAlive check is not needed. So we control if we need the keepAlive feature in BalanceInput or not.

We can introduce an UI hook here with following pattern:

`const {amount, hasEnoughFunds, BalanceInputComponent} = useBalanceInput(isKeepAliveNeeded: true/false)`

**Comment by ishrak-imam** [comment]
We will also need the unformatted existential deposit from the API to do the comparison.

File: src/ui/components/BalanceInput.tsx [hunk 10]
```diff
@@ -32,13 +33,15 @@ export function BalanceInput(props: PropTypes) {
...
-        error={!hasEnoughBalance}
+        error={Boolean(!hasEnoughBalance && amount)}
         mode="outlined"
         autoComplete="off"
         placeholder="Enter amount"
         keyboardType="decimal-pad"
         value={amount}
-        onFocus={() => setAmount('')}
+        onFocus={() => {
+          amount;
```
Comment (silva-fj) [inline src/ui/components/BalanceInput.tsx]: What's this?

File: src/ui/screens/SendFundScreen.tsx [hunk 11]
```diff
@@ -26,15 +32,31 @@ export function SendFundScreen({navigation, route}: Props) {
...
+  }, [toAddress, currentNetwork]);
+
+  const isEnteredBalanceValid = useMemo(() => {
+    const enteredBalance = stringToBn(amount) ?? BN_ZERO;
+    const isBalanceValid =
+      enteredBalance.gt(BN_ZERO) && enteredBalance.lt(formattedStringToBn(accountInfo?.balance.formattedFree));
+    return isBalanceValid;
+  }, [amount, accountInfo, stringToBn]);
+
+  const isSendDisabled = !isEnteredBalanceValid || !toAddress || !isToAddressValid;
```
Comment (silva-fj) [inline src/ui/screens/SendFundScreen.tsx]: You can remove `!toAddress` here, `isToAddressValid` will be false if `toAddress` is `undefined`

**Pull request merged**